### ISSUE TITLE
[Airsonic] Transcoding (13)

### ISSUE THREAD

**Issue by tesshucom** [tesshucom/jpsonic#1737]
## Problem description

This issue was ported from the Airsonic archive. The top node for these issues is #1725.

It does not necessarily aim to solve all problems. However, some points of view will be fed back as Jpsonic is improved in the future.

It also includes issues that have already been resolved by Jpsonic at this stage. These will be progressively organized by topic.

## System information

 * **Jpsonic version**: since v111.5.4

## New task is issued

 - airsonic/airsonic#1762
   - -> #1868

## Closed

Tracking of the following agenda is closed.

 - airsonic/airsonic#873
   - In Jpsonic, the video seek bar has been improved to work. (For non-transcoding)
 - airsonic/airsonic#472
 - airsonic/airsonic#1408
   - You should be able to use simple commands with ffmpeg. Better processing is technically possible, but redistribution of precompiled binaries is not legally possible. Users can get the source and compile locally. (explained in ffmpeg docs)
   - As mentioned often, FLAC is recommended. If you unintentionally find yourself in the swamp of vendor lock-in, you'll need to put in extra effort to get yourself out of it.
 - airsonic/airsonic#1783
   - It seems to be resolved. We do not plan to start improving our estimates now (because it would require significant changes). 
     - If anything, the estimation logic is supposed to be removed. #1741
   - We may revisit these after other critical issues are resolved. However, it should also be noted that unlike 10-20 years ago, the need for a feature called Transcoding has faded. If FLAC original data exists in the library, wireless-lossless viewing is possible with the configuration of Experia, UPnP app, and LDAC device. And such an environment is not particularly rare. In other words, the specs of peripheral devices have increased by several dozen times compared to 10 to 20 years ago, so there is room for consideration as to whether compression processing on the server side is the optimal solution overall. Even if there is some ambiguity, it is no longer the first area to spend time on. 'VBR? That reminds me, there was such a thing a long time ago ... ' Such a time will surely come someday
   -  By the way, personally, I haven't used VBR for music in over 20 years. Web discussions and heated streaks make a good impression, but be aware that people who see no need for them or have no idea about them tend not to participate.
 - airsonic/airsonic#1416
 - airsonic/airsonic#1698
   - It's a little maniacal・・・. I have some knowledge of their sound sources and their contemporaneous Roland/Yamaha standards and hardware. It is relatively difficult to reproduce in OSS software. (Pure, reproducibility of sound quality) I think it would be better to buy a dedicated device or a reissue sound card with compatible functions. If you use them and convert them to sound sources such as FLAC, you can avoid the situation where everything will die eventually.
 - airsonic/airsonic#525
   - I'm not positive about this feature. 
   - What is being discussed is technically possible, but has a higher load impact on the server. Where to handle it is fluid and not necessarily a requirement of the server. The user can resolve it even if the server doesn't support it.
     - A. You can also solve it when ripping in advance (in the first place, the idea of having the gain already resolved in the library)
     - B. Handle it on send (This issue)
     - C. Process when received
   - A is relatively classic and common. C is a feature that is likely to be present, especially for devices that use transmission standards that involve some compression technique. For example, modern high-performance wireless earphones, headphones, speakers, etc. Older people may know A, and younger people may avoid it by using C without knowing anything. Or maybe there is good software with good algorithms in A or C
   - The case where B is required is the case where neither A nor C can be secured. It's an edge case.
   - If such a function is installed, the next thing to come is the request for suppressors and effectors. In other words, it's an argument that puts one foot on the redevelopment of the wheel. Also, even if the feature was added, some people would **never want** to use it. It seems to be YAGNI even if it is an option. This is because there are costs for maintenance and version updates. Currently fffmpeg should be able to do quite a bit, so if you really need it, please incorporate it optionally.
 - airsonic/airsonic#830
   - > Would be nice to have a reset default button though !
     - Jpsonic has a default reset button.
     - If you have any other suggestions for the default transcoding, please open a new Issues
       - Note the version of ffmpeg. Conditions must be valid for both Docker's Alpine and Jammy versions of ffmpeg. (In other words, make sure it works on both) The reason is that Alpine is likely to use the latest version, and Jammy uses a somewhat generalized version.
 - airsonic/airsonic#1556
   - Jpsonic has done a major version upgrade of jaudiotagger, so this issue will not be reproduced. When upgrading, the problem of Airsonic is not inherited and fixed. I would rather recommend Jpsonic if you encounter a lot of glitches. If you have any strange problems with Jpsonic, then please open a new issue.
 - airsonic/airsonic#1744
   - It looks like a rudimentary permission setting issue. You can solve it by checking in order.
 - airsonic/airsonic#1789
   - > If i manually specify bitrate on the transcode command line, it works as expected (i tried with 32kpbs) When downsampling mp3 file from 320kbps, it works as expected.
     - As far as the phenomenon is concerned, it appears to be working as expected.
     - The specification of DSub is unknown, but if the Max Bitrate is not specified from the client, the Max Bitrate of the server-side player will be adopted. If the player's Max Bitrate is unspecified (infinite), 320k will be adopted for Flac to MP3 conversion. Because that is the limit available in MP3. This is a common specification for the Subsonic server family.
   - > My player setting is 160kbps max bitrate. This setting is applied correctly when playing 320k mp3 files (it outputs 160k mp3 files), but not with flac files (it outputs 320k mp3 files).
     - Sorry, I'm not clear on what this means. If you have 320k FLAC on the server and you request a 320k stream from the client, Airsonic may not be able to handle it. 320K MP3s are not always returned(Depends on DSub implementation).
       - This has been improved in Jpsonic. 
         - Subsonic/Airsonic : If the format is not specified by the client and the bit rate conditions are met, format conversion is not performed and Raw is sent.
         - Jpsonic : You can also specify a default format for each protocol. In other words, unlike the legacy, the use case of converting the format while keeping the bitrate is added. It is especially needed for UPnP and anonymous web players that cannot assign request parameters.
     - If you can provide more detailed information and specific problems, it may lead to improvement. If it does not work as expected, please provide us with the information and we will work with you until it is resolved.
       - On the other hand, at Jpsonic, question with insufficient information or no progress are considered statute of limitations and will be closed after a certain period of time. (It is the result of no progress for a certain period of time). If there is something that cannot be resolved, please open the new issue.
 - airsonic/airsonic#1825
   - Command chain processing time is longer than normal transcoding. Therefore, depending on the combination of server and network specifications and the implementation of the client application, it is quite possible that such a phenomenon will occur. There are currently no silver bullets. Conversely, if you make full use of sufficient specs, good apps, preload functions, etc., it may go well. This is what happens within the first few seconds of the song.
   - Players from well-known apps that appear to offer similar playback capabilities actually buffer in different ways. Varies by player.
     - Start playback with less data buffered first. Or, playback will start after sufficient buffering.
     - To start performance processing even if buffering does not end within a fixed time. Or notify timeout error.
     - Start playback in parallel with the buffering process. The player display is asynchronous, and the beginning of the song actually contains a few milliseconds of silence.
       - You'll think it's a silly design. These players actually exist, and they all use them without even realizing it. Human ears and eyes are unreliable. As a result, it looks like a very responsive music player.
   - Not limited to the beginning, there are various behaviors when transcoding is delayed during a performance.
     - Notify timeout
     - Go to the next song as if nothing happened
     - When the performance is momentarily interrupted, the standby is performed well, and when the data is sent, the performance is resumed from that point.
   - All the server side can do is prepare the environment for processing as quickly as possible, or omit redundant commands. There are client apps that are robust and those that are not. For high hurdle processing, a robust client app is recommended.
 - airsonic/airsonic#1581
   - Somewhere in Subsonic or Airsonic it is mentioned playing On the Fly, so perhaps this is off-topic for your use case.
   - I once thought about server-side caching and pre-transcoding. Now that doesn't seem like the way to go.
     - While transcoding remains an important feature, parts of the music market seem to be turning to bit-perfect thinking... When trying to distribute limited development resources well, I think other improvements tend to take precedence over transcoding functionality. Except for defects.
